### DISCUSSION TITLE
Is it allowed to run Vercel behind a Reverse Proxy?

### DISCUSSION THREAD

**Discussion by gopeter** [vercel/vercel#5769] [Help] [answered]
We're currently building a proof-of-concept for a customer, where we need to proxy a lot of backend services together in an own Nginx setup, which sits behind a Cloudflare CDN.

I am aware that hosting on Vercel also provides a lot of CDN functions, but we need some of the security and firewall features that Cloudflare provides.

So what we are want to do is basicallyy this:

```
server {
  listen 80;

  location /backend1 {
    proxy_pass https://backend1/;
  }

  location /backend2 {
    proxy_pass https://backend2/;
  }

  location / {
    proxy_pass https://my-app.vercel.app/;
    proxy_redirect off;
    proxy_set_header X-Real-IP $remote_addr;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
  }
}
```

This setup works fine, but I wonder if there are any downsides and if it's allowed? I looked at the terms of service, but didn't found any hints or warnings about running Vercel apps behind a proxy.

**Comment by paulogdm** [comment, marked as answer]
Hello 👋 ,

Glad to hear you are considering Vercel for a future project! I will summarize my thoughts below 👇 

# Short Answer

Using a proxy in front of Vercel **is not recommended**.
EDIT: Please refer to https://vercel.com/support/articles/can-i-use-a-proxy-on-top-of-my-vercel-deployment

# Long Answer

You can always use something on top of Vercel, as long as you have full control over the process and deep knowledge about your setup. We optimized the platform in several ways to make sure it is **fast** globally. When you introduce your proxy, it will fail to perform to your audience if it is not distributed. It can become a bottleneck and source for all sorts of problems. A few things to consider:

- When a new deployment happens, how will you purge the cache of any external services? Remember, Vercel does this automatically for you.
- How can you guarantee your proxy is not introducing an additional slow hop to your user's requests?
- On Vercel, you must add the production domain to your project, and it should be configured properly. In your case, it seems the DNS configuration wouldn't respect that.
- **SEO**: We would not send an "X-Robots-Tag: noindex" automatically. If you want to proceed, make sure that is the case, so only the domain.

At the end of the day, your setup would be so complex that the meaning of Vercel for your project would diminish. Maintaining this type of setup is not a good idea overall. Please check if you can do the following instead:

- **Incremental Adoption**: https://nextjs.org/blog/incremental-adoption. Go to Vercel first, then proxy.
-  Use a subdomain for the backend: It is common to reserve `api.example.com`. Wouldn't that be better to isolate both systems?

**Reply by gopeter** [reply]
Dear @paulogdm,

thank you very much for the detailed answer. I'm going to dive a little bit deeper into our needs to better understand what we're trying to achieve. Maybe that's helpful for others too.

The bad news first: using subdomains would solve all of the issues, but it's not possible in our case. The customer want to have a seamless integration of all services. Since I can't give more details, I try to give a more abstract example: if a user visits a webshop, adds something to the cart and then proceeds to the checkout, the domain (and the whole user experience) should remain the same, even if /products, /cart and /checkout are all seperate processes/services, which we glue together in our Nginx proxy. Some of the services are provided by third party vendors, some of them are own services running in their own containers. 

That's why we have to build a solution which runs all of this under one domain. And that's where the reverse proxy comes in to provide (parts of) the frontend via Vercel. But why should we run a CDN which serves a Proxy which runs a proxy to another CDN? Lot of round trips and wasted resources. **But:** running the frontend on Vercel gives us the ability to trigger new builds with a webhook. If the customer make changes in the backend, we can trigger the webhook and a few minutes later, the frontend serves our static generated pages. 

If we would run Next.js by ourself, we would need to spawn at least two instances of it, running behind a load balancer (for example on AWS ELB). But then we are not able to trigger new builds, since we don't have "access" to the instances behind the load balancer and therefore can't make sure that every instance has the new/updated content. The same applies for ISG. 

So, for now, we have to options:
- Proxy Vercel: gives us the abilitiy to trigger new builds, but adds complexity and potential performance ingressions
- Running Next.js on our own on AWS ELB (for example) and lose the ability to use SSG/ISG

Maybe there is a third option: [running Next.js on AWS Lambda@Edge](https://github.com/serverless-nextjs/serverless-next.js), but I have to dive deeper into this.

Some more thoughts on your points:

> On Vercel, you must add the production domain to your project, and it should be configured properly. In your case, it seems the DNS configuration wouldn't respect that.

We did not added any domains to Vercel and were able to proxy my-app.vercel.app successfully with Nginx under our own domain my-app.com.

> SEO: We would not send an "X-Robots-Tag: noindex" automatically. If you want to proceed, make sure that is the case, so only the domain.

Yes, in fact, we would have to do this by ourselfs.

**Reply by paulogdm** [reply]
> The bad news first: using subdomains would solve all of the issues, but it's not possible in our case. The customer want to have a seamless integration of all services. Since I can't give more details, I try to give a more abstract example: if a user visits a webshop, adds something to the cart and then proceeds to the checkout, the domain (and the whole user experience) should remain the same, even if /products, /cart and /checkout are all seperate processes/services, which we glue together in our Nginx proxy. Some of the services are provided by third party vendors, some of them are own services running in their own containers.

Using incremental adoption could be a good call. Check https://nextjs.org/blog/incremental-adoption. Use Vercel to proxy to your NGINX instead.

> But: running the frontend on Vercel gives us the ability to trigger new builds with a webhook. If the customer make changes in the backend, we can trigger the webhook and a few minutes later, the frontend serves our static generated pages.

If the code does not change, only the data, ISG/SSR + caching headers should be used instead. I would recommend that you avoid webhook calls and rebuilds if data in the CMS changes. The main purpose of ISG is to make those workflows outdated.

---

Overall, I would recommend that you re-evaluate Vercel. We want to make your development experience as smooth as possible. If your use-case requires too many workarounds, it defeats the purpose of using Vercel. At the end of the day, you would spend more time caring about the infrastructure, hunting weird bugs due to the unusual setup, and maintaining the structure instead of focusing on your product.

**Reply by gopeter** [reply]
> Using incremental adoption could be a good call. Check https://nextjs.org/blog/incremental-adoption. Use Vercel to proxy to your NGINX instead.

We also considered that, but we need more "control" over the CDN, to configure Firewalls, WAF, and so one.

> If the code does not change, only the data, ISG/SSR + caching headers should be used instead. I would recommend that you avoid webhook calls and rebuilds if data in the CMS changes. The main purpose of ISG is to make those workflows outdated.

ISG is great, but we have pages where we can't rely on it. If a customer gets an outdated version of shop page, he would be able to put things into the basket which are not available any more. And since we are running everything on multiple instances behind a load balancer, we can't force ISG by just `curl`ing it once (at the load balancer level).

> Overall, I would recommend that you re-evaluate Vercel. We want to make your development experience as smooth as possible. If your use-case requires too many workarounds, it defeats the purpose of using Vercel. At the end of the day, you would spend more time caring about the infrastructure, hunting weird bugs due to the unusual setup, and maintaining the structure instead of focusing on your product.

Yes, I think Vercel isn't for us. I've played around with the `serverless`-template, which I mentioned above, and I think that this fits into our whole stack.

Anyway, thank you very much for your thoughts and detailed answers!

**Reply by paulogdm** [reply]
> ISG is great, but we have pages where we can't rely on it. If a customer gets an outdated version of shop page, he would be able to put things into the basket which are not available any more. And since we are running everything on multiple instances behind a load balancer, we can't force ISG by just curling it once (at the load balancer level).

We solved this in our commerce example. Basically, you can introduce "revalidation" on the client with SWR, for example. `stale-while-revalidate` + an API route + client-side bouncer can give you a very good experience. Using SSR so you can "always receive the most up-to-date" information is indeed a common approach, but it hurts the speed of your website. The most successful Vercel-deployed e-commerces are using ISG or SSR + `stale-while-revalidate`.

**Reply by joffreyBerrier** [reply]
Hi @paulogdm any news about that? It's 1 year after :')

**Comment by pratik-dani** [comment]
> We're currently building a proof-of-concept for a customer, where we need to proxy a lot of backend services together in an own Nginx setup, which sits behind a Cloudflare CDN.
> 
> I am aware that hosting on Vercel also provides a lot of CDN functions, but we need some of the security and firewall features that Cloudflare provides.
> 
> So what we are want to do is basicallyy this:
> 
> ```
> server {
>   listen 80;
> 
>   location /backend1 {
>     proxy_pass https://backend1/;
>   }
> 
>   location /backend2 {
>     proxy_pass https://backend2/;
>   }
> 
>   location / {
>     proxy_pass https://my-app.vercel.app/;
>     proxy_redirect off;
>     proxy_set_header X-Real-IP $remote_addr;
>     proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
>   }
> }
> ```
> 
> This setup works fine, but I wonder if there are any downsides and if it's allowed? I looked at the terms of service, but didn't found any hints or warnings about running Vercel apps behind a proxy.

Hey @gopeter, What config did you use to proxy your vercel app on your domain (yourdomain.com/vercel). I am trying this but I am getting errors.

**Reply by joffreyBerrier** [reply]
Hi @devildani any news about this ?
